homeassistant/components/homekit/__init__.py

```diff
@@ -3,154 +3,197 @@
+        vol.Optional(CONF_ENTITY_CONFIG, default={}): validate_entity_config,
     })
 }, extra=vol.ALLOW_EXTRA)
 
 
-@asyncio.coroutine
-def async_setup(hass, config):
+async def async_setup(hass, config):
     """Setup the HomeKit component."""
-    _LOGGER.debug("Begin setup HomeKit")
+    _LOGGER.debug('Begin setup HomeKit')
 
     conf = config[DOMAIN]
-    port = conf.get(CONF_PORT)
-    pin = str.encode(conf.get(CONF_PIN_CODE))
+    port = conf[CONF_PORT]
+    auto_start = conf[CONF_AUTO_START]
+    entity_filter = conf[CONF_FILTER]
+    entity_config = conf[CONF_ENTITY_CONFIG]
 
-    homekit = HomeKit(hass, port)
-    homekit.setup_bridge(pin)
+    homekit = HomeKit(hass, port, entity_filter, entity_config)
+    homekit.setup()
 
-    hass.bus.async_listen_once(
-        EVENT_HOMEASSISTANT_START, homekit.start_driver)
-    return True
+    if auto_start:
+        hass.bus.async_listen_once(EVENT_HOMEASSISTANT_START, homekit.start)
+        return True
+
+    def handle_homekit_service_start(service):
+        """Handle start HomeKit service call."""
+        if homekit.started:
+            _LOGGER.warning('HomeKit is already running')
+            return
+        homekit.start()
 
+    hass.services.async_register(DOMAIN, SERVICE_HOMEKIT_START,
+                                 handle_homekit_service_start)
 
-def import_types():
-    """Import all types from files in the HomeKit directory."""
-    _LOGGER.debug("Import type files.")
-    # pylint: disable=unused-variable
-    from . import (  # noqa F401
-        covers, security_systems, sensors, switches, thermostats)
+    return True
 
 
-def get_accessory(hass, state):
+def get_accessory(hass, state, aid, config):
     """Take state and return an accessory object if supported."""
+    _LOGGER.debug('%s: <aid=%d config=%s>')
+    if not aid:
+        _LOGGER.warning('The entitiy "%s" is not supported, since it '
+                        'generates an invalid aid, please change it.',
+                        state.entity_id)
+        return None
+
     if state.domain == 'sensor':
         unit = state.attributes.get(ATTR_UNIT_OF_MEASUREMENT)
         if unit == TEMP_CELSIUS or unit == TEMP_FAHRENHEIT:
-            _LOGGER.debug("Add \"%s\" as \"%s\"",
+            _LOGGER.debug('Add "%s" as "%s"',
                           state.entity_id, 'TemperatureSensor')
             return TYPES['TemperatureSensor'](hass, state.entity_id,
-                                              state.name)
+                                              state.name, aid=aid)
 
     elif state.domain == 'cover':
         # Only add covers that support set_cover_position
-        if state.attributes.get(ATTR_SUPPORTED_FEATURES) & 4:
-            _LOGGER.debug("Add \"%s\" as \"%s\"",
-                          state.entity_id, 'Window')
-            return TYPES['Window'](hass, state.entity_id, state.name)
+        if state.attributes.get(ATTR_SUPPORTED_FEATURES, 0) & 4:
```

Comment:
I would prefer if you use `cover.SUPPORT_SET_POSITION` so that it keeps the code readable.